### PR TITLE
(CDPE-6584) skip PaxHeaders.X entries

### PR DESCRIPTION
**Pull request by davidmalloncares**
Skip PaxHeaders.X entries as they interfere with logic to work out long filenames.

This is related to a customer issue (CDPE-6584) where the cd4pe_job.tar.gz file gets extracted but some file names are truncated, e.g. 

`cd4pe_job/repo/spec/classes/controls/rhel_8/v3_0_0/initial_setup/configure_additional_process_harden`

instead of the desired:

`cd4pe_job/repo/spec/classes/controls/rhel_8/v3_0_0/initial_setup/configure_additional_process_hardening_spec.rb`

I don't know Ruby or Tar but this change allows the tar file in question to be extracted with the full file names intact, its a starting point for a conversation at least.

I worked out this potential fix with a mixture of debug messages and reading over the code - it looked like the code expected 2 tar entries to cover files with a long filename - on the first entry it encounters the 'LongLink' situation and correctly works out the full filename then proceeds to the next iteration where it uses the correct, full filename worked out on the previous iteration to create the file with the correct name. I saw that there was an extra 'PaxHeader' entry interfering with this logic and causing the issue.  

Added unit test for new PAX header functionality. See images of failing test without fix, and passing test with fix.

**Without code fix**

Screenshot 2024-03-25 at 17 13 28

**With code fix** 

Screenshot 2024-03-25 at 17 14 17

### CODE REVIEW

File: tasks/run_cd4pe_job.rb [hunk 1]
```diff
@@ -53,6 +53,10 @@ def self.preserve_symlink(entry, dest)
     Gem::Package::TarReader.new( Zlib::GzipReader.open zipped_file_path ) do |tar|
       dest = nil
       tar.each do |entry|
+        # skip 'PaxHeaders.X' entries as they interfere with the TAR_LONGLINK logic
+        if !(entry.directory? || entry.file? || entry.header.typeflag == SYMLINK_SYMBOL) && entry.full_name != TAR_LONGLINK
```
Comment (sean-mckenna) [inline tasks/run_cd4pe_job.rb:57]: I noticed when I printed the `entry.header.typeflag` for a PaxHeader it was always "x". Perhaps we can change this check to explicitly check for this value?

We can add a PAX_HEADER variable alongside TAR_LONGLINK and SYMLINK_SYMBOL above like this

```ruby
        PAX_HEADER = 'x'.freeze
```
Then change the if statement to 

```ruby

        if entry.header.typeflag == PAX_HEADER
          next
        end
```

Comment (davidmalloncares) [reply tasks/run_cd4pe_job.rb]: Yeah thats def much nicer

File: tasks/run_cd4pe_job.rb [hunk 2]
```diff
@@ -29,6 +29,7 @@ class GZipHelper
   # static class to decompress tar.gz files
   TAR_LONGLINK = '././@LongLink'.freeze
   SYMLINK_SYMBOL = '2'.freeze
+  PAX_HEADER = 'x'.freeze
```
Comment (Magisus) [inline tasks/run_cd4pe_job.rb:32]: It sounds like it is also possible to have a pax header with a 'g' typeflag (which affects all files in the archive, whereas 'x' just affects the next file). See https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html#tag_20_92_13_02. Do we need to account for that too?

**Review by Magisus — COMMENTED** [review]
Do you have any resources you could link to about how you figured this out? Curious in what way these headers interfere with Longlink. 

Also do we know why it truncates at 38? The comment about Longlink applying to paths over 100 characters seems consistent with other things I've read about it.

Lastly, is this something we can write a test for? That could be really helpful to show an example of when this code comes into play. I think we have other tests that exercise the longlink behavior already.

File: tasks/run_cd4pe_job.rb (hunk 2 above)
Comment (davidmalloncares) [reply tasks/run_cd4pe_job.rb:32]: We haven't seen that in the example tar file from the customer so I think we are ok to proceed with the change as is until there is evidence that a 'g' type flag is causing an issue.

**Review by Magisus — APPROVED** [review]

**Pull request merged**